### PR TITLE
Refreshes to clear heap used by version map should not be visible to users

### LINKED ISSUE #15768: Use separate searchers for "search visibility" vs "move indexing buffer to disk"
Spinoff from #14121...

Today, when ES detects it's using too much heap vs the configured indexing buffer (default 10% of JVM heap) it opens a new searcher to force Lucene to move the bytes to disk, clear version map, etc.

But this has the unexpected side effect of making newly indexed/deleted documents visible to future searches, which is not nice for users who are trying to prevent that, e.g. #3593.

As @uschindler suggested in that issue, I think ES should have two separate searchers from the engine: one for search visibility, only ever refreshed according to the user's wishes, and another, used internally for freeing up heap, version map lookups, etc.  Lucene will be efficient about this, sharing segment readers across those two searchers.

I haven't started on this (need to finish #14121 first!) so if someone wants to take it, please feel free!

### PR DESCRIPTION
**Pull request by mikemccand**
This change makes it more predictable/controllable to users when recently indexed documents become visible for searches via refresh by making version map use a private searcher so that when it needs to free up heap, its refreshes won't be visible to users.

Note that this also means flush no longer also refreshes, so I had to fix some tests that were relying on this.

Note that as @bleskes explained on #157658 this won't be perfect, e.g. when shards relocate there is a user-visible refresh automatically done.  Progress not perfection...

Closes #15768

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -285,14 +296,15 @@ protected void recoverFromTranslog(EngineConfig engineConfig, Translog.TranslogG
...
+        DirectoryReader directoryReader = null;
         SearcherManager searcherManager = null;
         try {
             try {
-                final DirectoryReader directoryReader = ElasticsearchDirectoryReader.wrap(DirectoryReader.open(indexWriter, true), shardId);
+                directoryReader = ElasticsearchDirectoryReader.wrap(DirectoryReader.open(indexWriter, true), shardId);
                 searcherManager = new SearcherManager(directoryReader, searcherFactory);
-                lastCommittedSegmentInfos = readLastCommittedSegmentInfos(searcherManager, store);
-                success = true;
+                if (lastCommittedSegmentInfos == null) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: This feels a bit weird. Maybe we should separate this into two methods:
1. create the reader and read the lastComittedInfos
2. create a searcherManger based on the given reader. If one needs a dedicated directoryReader per searcher manager (not sure of lucene semantics), can we maybe call  `lastCommittedSegmentInfos = readLastCommittedSegmentInfos(searcherManager, store);` once in the constructor?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -304,8 +316,8 @@ private SearcherManager createSearcherManager() throws EngineException {
                 throw new EngineCreationFailureException(shardId, "failed to open reader on writer", e);
             }
         } finally {
-            if (success == false) { // release everything we created on a failure
-                IOUtils.closeWhileHandlingException(searcherManager, indexWriter);
+            if (searcherManager == null) { // release everything we created on a failure
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think this may a leak an unclosed searchManager if the second one runs into trouble (than only the second one will be closed). Maybe we should deal with all of that the constructor? we already have a try catch there?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -330,13 +342,19 @@ public GetResult get(Get get, Function<String, Searcher> searcherFactory) throws
                         return new GetResult(true, versionValue.version(), op.getSource());
                     }
                 }
-            }
 
-            // no version, get the version from the index, we know that we refresh on flush
-            return getFromSearcher(get, searcherFactory);
+                return getFromSearcher(get, this::acquireInternalSearcher);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: this seems to bypass the IndexSearcherWrapper defined in IndexShard (and passed via the searchFactory)

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 4]
```diff
@@ -495,6 +513,8 @@ public void refresh(String source) throws EngineException {
         try (ReleasableLock lock = readLock.acquire()) {
             ensureOpen();
             searcherManager.maybeRefreshBlocking();
+            IndexSearcher s = searcherManager.acquire();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: why do we need to acquire a searcher here?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 5]
```diff
@@ -505,7 +525,29 @@ public void refresh(String source) throws EngineException {
             throw new RefreshFailedEngineException(shardId, t);
         }
 
-        // TODO: maybe we should just put a scheduled job in threadPool?
+        mergeScheduler.refreshConfig();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Is this a separate bug you found?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 6]
```diff
@@ -505,7 +525,29 @@ public void refresh(String source) throws EngineException {
...
+
+        // we must also refresh our internal searcher, so that a subsequent real-time get (which uses internal searcher) comes from the
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: correct me if I'm wrong, but this is about trimming the version map, right?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 7]
```diff
@@ -505,7 +525,29 @@ public void refresh(String source) throws EngineException {
...
+        // index and not xlog:
+        refreshInternal();
+    }
+
+    private void refreshInternal() throws EngineException {
+        // we obtain a read lock here, since we don't want a flush to happen while we are refreshing
+        // since it flushes the index as well (though, in terms of concurrency, we are allowed to do it)
+        try (ReleasableLock lock = readLock.acquire()) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: do we want to unify this code with the refresh method by making this get a manager to work on? (+ a string description for failures)

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 5 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: it's done in refreshInternal as well.. so I'm probably missing something w.r.t. lucene.  Looking to learn ;)

**Comment by bleskes** [comment]
Left some comments/questions here and there. While I may need similar functionality for the seq_no work (later on, not sure), I still have my doubts whether the change is worth the added complexity here. The engine is complex enough as it is and since this doesn't really give us 100% guarantees (as discussed) I feel like we are maybe sending people the wrong message here. Uwe wanted to control doc visibility so he can build on that and have some logic in his application. Now it will seem to work but sometimes wont (in subtle ways). I'm leaning towards saying we should be more upfornt about it which will also allow us to avoid some complexity.

Last, I'm fine with not refreshing after an internal flush, but I think we should if someone calls the explicit flush API. If a user calls flush (i.e., put everything into lucene and trim the translog) I would expect the docs (now part of lucene guaranteed) to be part of searchers. I understand that the underlying semantics can (and are) different but I think it will be confusing to users.  I wonder how other people feel.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: This is a great idea (single `DirectoryReader` on init, two searcher managers).

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 2 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think we are actually OK here: the finally clause in the ctor would close the first searcherManager.  I agree it's confusing ...

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 4 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Good catch, no reason!

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 5 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: This is pre-existing, and I don't know why refresh means it's a good time to ask merge scheduler to pull any settings changes.  We already call this in `onSettingsChanged`... I suspect it's not really necessary :)

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 6 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: There were RT get tests that were angry, because apparently an "RT get from xlog" can sometimes fail (if there were mapping changes maybe, not sure?) where "RT get from the index" would succeed.  So if an explicit refresh is invoked I need to refresh both internal and external readers.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 7 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: ++

**Comment by mikemccand** [comment]
> I still have my doubts whether the change is worth the added complexity here.

I don't think there's so much added complexity in this change.  Yes, we have two `SearcherManagers`, but they share a single reader pool inside `IndexWriter` so performance should be very nearly the same except when the two managers diverge substantially.

I really don't like that internal implementation details of ES's heap management randomly impact user visibility of recent changes.  I think it's important to minimize such leaky abstractions.  Solr already uses a private reader for its realtime lookups, though presumably is also not "perfect" when a shard moves.

I think if ES did segment (file) based replication instead of document based, such goals would be more natural since each node would have exactly the same point-in-time searcher versions visible (yeah, I know, that'd be a huge change, lots of tradeoffs).

I think for this change, we could maybe also use Lucene's `SearcherLifetimeManager` instead of two `SearcherManagers` ... it could be simpler since it'd mean we just have to hold the private and public searchers, pulled from a single `SearcherManager`.

Finally, if this really is a non-goal, i.e. the use case @uschindler asks about can't and won't be supported and is unimportant, then I agree any iota of added complexity is not worth it, and we should not pursue it.

**Comment by bleskes** [comment]
> I don't think there's so much added complexity in this change. ... I really don't like that internal implementation details of ES's heap management randomly impact user visibility of recent changes. I think it's important to minimize such leaky abstractions.

Fair enough. This is a judgment call. I don't feel too strongly about it. I wonder what other think.

**Comment by uschindler** [comment]
Hi,

the request stated as "my request" was one of my customer's requests (back at that time). If you check your records you might be able to find out which of Elastics customer's this might be was responsible for this request. The current company's politics in supporting your valuable supporters in finding you  even new customers (like offering free tickets to a conference over the ocean without also offering free flights to the conference or offering free hotel rooms) does not really help to make me spending any more time in this issue.

So my +/-0 is here :-)

**Comment by uschindler** [comment]
To come back to issue:

Basically I think what we need:

Lucene out of box provides the following functionality:
- you can do a commit
- you can delete tons of documents (delete by query which was problem here)
- then index new documents
- commit again

What Lucene provides is that the user sees this as one transaction! This is requested quite often by people when they want to replace some of their data, but cannot do that by deleting and updating single documents. They often can only remove all documents of one category and replace them by new ones. With delete-by-query refreshing automatically (like it did/does) this gets impossible!

Maybe we can add something else: How about adding an API to temporarily disable everything of Elasticsearch docing in background - also replication. Then do your updates (e.g. with 2nd searcher) and then call another API to reenable the whole system?

Basically, I and many other just want to have back the basic and trivial commit/snapshot functionality of Lucene, hwich works with Lucene and Solr, but not Elasticsearch.

**Comment by uschindler** [comment]
Basically the API is there, it is called "disable refresh" and "enable refresh", but the stuff to be fixed here breaks it.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 5 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: no you can remove this - it's called when needed in onSettingsChanged

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 8]
```diff
@@ -161,15 +167,20 @@ public InternalEngine(EngineConfig engineConfig, boolean skipInitialTranslogReco
                 }
             }
             this.translog = translog;
-            manager = createSearcherManager();
-            this.searcherManager = manager;
-            this.versionMap.setManager(searcherManager);
+            searcherManager = createSearcherManager();
+            this.searcherManager = searcherManager;
+
+            internalSearcherManager = createSearcherManager();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think the internalSearcherManager doesn't need the same searcher factory as we use for the external one since it doesn't need to do warming etc? I think it should be very lean?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: ++

**Comment by s1monw** [comment]
I think this change is fine and IMO the right thing to do. 

@uschindler the only reason why we could call refresh is a relocation of a shard and you can already disable this per index if you want. once this change is in

**Comment by uschindler** [comment]
OK!

**Comment by dakrone** [comment]
@mikemccand what's the word with this PR? Is it waiting on anything still?

**Comment by mikemccand** [comment]
Sorry for the delays here ... this was somewhat controversial, and doesn't fully solve the @uschindler example (because relocations can still bring changes live) so I didn't push this change.  I'll remove 5.0.0 tag and put discuss on.

**Comment by bleskes** [comment]
we discussed this at fix it friday and the general consensus was that the requested feature (full control of refresh cycles for full control of data visibility) is a high hanging fruit and we rather not start embarking on that right now. @mikemccand are you OK with closing this?

**Comment by mikemccand** [comment]
Sure, I'll close it.
